### PR TITLE
apple: Add WSI support for tests on MacOS

### LINKED ISSUE #6412: Add WSI support for tests on MacOS
(from https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/6407)

right now a lot of WSI tests are blacklisted for MacOS or are just going

```
    if (!InitSwapchain()) {
        GTEST_SKIP() << "Cannot create surface or swapchain";
    }
```

but really we just need to add proper fixes to the `AddSurfaceExtension`, `CreateSurface`, etc helpers

### PR DESCRIPTION
**Pull request by juan-lunarg**
Required adding objective-c++ code into the test codebase to properly aquire the MetalLayer

closes #6412

<details>

```
Note: Google Test filter = *PositiveWsi*
[==========] Running 16 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 16 tests from PositiveWsi
[ RUN      ] PositiveWsi.UseDestroyedSwapchain
Driver Name = MoltenVK
Driver Info = 1.2.6
[       OK ] PositiveWsi.UseDestroyedSwapchain (2840 ms)
[ RUN      ] PositiveWsi.CreateWaylandSurface
/Users/juanramos/projects/vvl/tests/unit/wsi_positive.cpp:23: Skipped
test not supported on platform

[  SKIPPED ] PositiveWsi.CreateWaylandSurface (1 ms)
[ RUN      ] PositiveWsi.CreateXcbSurface
/Users/juanramos/projects/vvl/tests/unit/wsi_positive.cpp:94: Skipped
test not supported on platform

[  SKIPPED ] PositiveWsi.CreateXcbSurface (1 ms)
[ RUN      ] PositiveWsi.CreateX11Surface
/Users/juanramos/projects/vvl/tests/unit/wsi_positive.cpp:129: Skipped
test not supported on platform

[  SKIPPED ] PositiveWsi.CreateX11Surface (0 ms)
[ RUN      ] PositiveWsi.CmdCopySwapchainImage
[       OK ] PositiveWsi.CmdCopySwapchainImage (49 ms)
[ RUN      ] PositiveWsi.TransferImageToSwapchainDeviceGroup
/Users/juanramos/projects/vvl/tests/unit/wsi_positive.cpp:310: Skipped
minImageExtent is not large enough

[  SKIPPED ] PositiveWsi.TransferImageToSwapchainDeviceGroup (42 ms)
[ RUN      ] PositiveWsi.SwapchainAcquireImageAndPresent
[       OK ] PositiveWsi.SwapchainAcquireImageAndPresent (36 ms)
[ RUN      ] PositiveWsi.SwapchainAcquireImageAndWaitForFence
[       OK ] PositiveWsi.SwapchainAcquireImageAndWaitForFence (34 ms)
[ RUN      ] PositiveWsi.WaitForAcquireFenceAndIgnoreSemaphore
[       OK ] PositiveWsi.WaitForAcquireFenceAndIgnoreSemaphore (36 ms)
[ RUN      ] PositiveWsi.WaitForAcquireSemaphoreAndIgnoreFence
[       OK ] PositiveWsi.WaitForAcquireSemaphoreAndIgnoreFence (36 ms)
[ RUN      ] PositiveWsi.SwapchainImageLayout
[       OK ] PositiveWsi.SwapchainImageLayout (37 ms)
[ RUN      ] PositiveWsi.SwapchainPresentShared
/Users/juanramos/projects/vvl/tests/framework/render.cpp:349: Skipped
VK_KHR_shared_presentable_image not supported

[  SKIPPED ] PositiveWsi.SwapchainPresentShared (22 ms)
[ RUN      ] PositiveWsi.CreateSurface
[       OK ] PositiveWsi.CreateSurface (34 ms)
[ RUN      ] PositiveWsi.SwapchainImageFormatProps
[       OK ] PositiveWsi.SwapchainImageFormatProps (47 ms)
[ RUN      ] PositiveWsi.SwapchainExclusiveModeQueueFamilyPropertiesReferences
[       OK ] PositiveWsi.SwapchainExclusiveModeQueueFamilyPropertiesReferences (33 ms)
[ DISABLED ] PositiveWsi.DISABLED_ProtectedSwapchainImageColorAttachment
[ RUN      ] PositiveWsi.CreateSwapchainWithPresentModeInfo
[       OK ] PositiveWsi.CreateSwapchainWithPresentModeInfo (34 ms)
[----------] 16 tests from PositiveWsi (3289 ms total)

[----------] Global test environment tear-down
[==========] 16 tests from 1 test suite ran. (3316 ms total)
[  PASSED  ] 11 tests.
[  SKIPPED ] 5 tests, listed below:
[  SKIPPED ] PositiveWsi.CreateWaylandSurface
[  SKIPPED ] PositiveWsi.CreateXcbSurface
[  SKIPPED ] PositiveWsi.CreateX11Surface
[  SKIPPED ] PositiveWsi.TransferImageToSwapchainDeviceGroup
[  SKIPPED ] PositiveWsi.SwapchainPresentShared

  YOU HAVE 1 DISABLED TEST
```

</details>

### CODE REVIEW

**Comment by juan-lunarg** [comment]
Getting a failure on MacOS CI I cannot reproduce locally:

```
[ RUN      ] VkArmBestPracticesLayerTest.PresentModeTest
Driver Name = MoltenVK
Driver Info = 1.2.5
/Users/lunarg/.jenkins/vz0/Release64/Vulkan-ValidationLayers/tests/unit/arm_best_practices.cpp:607: Failure
Expected equality of these values:
  VK_ERROR_VALIDATION_FAILED_EXT
  vk::CreateSwapchainKHR(device(), &swapchain_create_info, nullptr, &m_swapchain)
    Which is: VK_SUCCESS
```

All `VkArmBestPracticesLayerTest` tests pass with MoltenVK 1.2.6 on my machine.

File: tests/unit/arm_best_practices.cpp [hunk 1]
```diff
@@ -604,11 +602,16 @@ TEST_F(VkArmBestPracticesLayerTest, PresentModeTest) {
...
-    ASSERT_EQ(VK_ERROR_VALIDATION_FAILED_EXT, vk::CreateSwapchainKHR(device(), &swapchain_create_info, nullptr, &m_swapchain));
-    m_errorMonitor->VerifyFound();
+    {
+        m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_WARNING_BIT_EXT,
+                                             "UNASSIGNED-BestPractices-vkCreateSwapchainKHR-swapchain-presentmode-not-fifo");
+
+        const auto err = vk::CreateSwapchainKHR(device(), &swapchain_create_info, nullptr, &m_swapchain);
+
+        ASSERT_TRUE(err == VK_ERROR_VALIDATION_FAILED_EXT) << string_VkResult(err);
+        m_errorMonitor->VerifyFound();
```
Comment (juan-lunarg) [inline tests/unit/arm_best_practices.cpp:612]: I'm unable to reproduce the failure even with 1.2.5 of MoltenVK. So I'm at least going to try and figure out the return value.

**Review by spencer-lunarg — APPROVED** [review]
This is great, not sure how any the objective C stuff actually works, but its nicely isolated

File: tests/unit/arm_best_practices.cpp (hunk 1 above)
Comment (juan-lunarg) [reply tests/unit/arm_best_practices.cpp:612]: CI is returning VK_SUCCESS but locally it's returning VK_ERROR_VALIDATION_FAILED_EXT and the version of molten vk doesn't appear to be the discrepancy.

Comment (spencer-lunarg) [reply tests/unit/arm_best_practices.cpp:612]: there is a real non-zero chance this test is written bad and we don't have all the WSI checks in to know it

Comment (spencer-lunarg) [reply tests/unit/arm_best_practices.cpp:612]: (aka, happy to blacklist this test and/or skip if using portability)

Comment (juan-lunarg) [reply tests/unit/arm_best_practices.cpp:612]: No the test is fine. CI isn't using the correct validation layer.

**Comment by juan-lunarg** [comment]
Will re-run on Jenkins once CI has been fixed.

**Pull request merged**